src/functions.rs

```diff
@@ -2919,6 +2919,75 @@ pub fn initialize_terminated<'a>(functions: &'a FunctionMap<'a>) -> (Vec<String>
     (term_ret_vec, nonterm_ret_vec)
 }
 
+// Helper function to substitue class name if needed
+fn cil_name_helper(
+    call: FuncCall,
+    func_info: &FunctionInfo,
+    function_map: &FunctionMap,
+) -> Result<String, CascadeErrors> {
+    let mut original_cil_name = call.get_cil_name();
+
+    // Deal with casting
+    if let Some(cast_class) = call.cast_name {
+        // This should be mostly safe since we have already looked at castables in prevalidate
+        original_cil_name = get_cil_name(Some(&cast_class), &call.name);
+    }
+    // If we are calling something with this or None it must be in the same class
+    // so hand place the class name
+    else if call.class_name.as_ref() == Some(&CascadeString::from("this"))
+        || call.class_name.is_none()
+    {
+        if let FunctionClass::Type(class) = func_info.class {
+            original_cil_name = get_cil_name(Some(&class.name), &call.name);
+        } else {
+            return Err(CascadeErrors::from(
+                ErrorItem::make_compile_or_internal_error(
+                    "Could not determine class for 'this.' function call",
+                    Some(func_info.declaration_file),
+                    call.get_name_range(),
+                    "Perhaps you meant to place the function in a resource or domain?",
+                ),
+            ));
+        }
+    }
+    // Finally deal with alias
+    if let Some(call_func_info) = function_map.get(&original_cil_name) {
+        original_cil_name = call_func_info.get_cil_name();
+    }
+
+    Ok(original_cil_name)
+}
+
+// Search through the non terminating functions to find loops
+fn find_recursion_loop(
+    func: &str,
+    function_map: &FunctionMap,
+    terminated_list: &Vec<String>,
+    visited: &mut Vec<String>,
+) -> Result<Vec<String>, CascadeErrors> {
+    visited.push(func.to_string());
+    if let Some(function_info) = function_map.get(func) {
+        let func_calls = get_all_func_calls(function_info.original_body.to_vec());
+        for call in func_calls {
+            if call.check_builtin().is_some() {
+                continue;
+            }
+            let call_cil_name = cil_name_helper(call, function_info, function_map)?;
+
+            if terminated_list.contains(&call_cil_name) {
+                continue;
+            } else if visited.contains(&call_cil_name) {
+                break;
+            } else if function_map.get(&call_cil_name).is_none() {
```

Comment:
This is the thing we discussed last week where the function doesn't exist, but we don't know that yet, because this is before validate_functions(), right?  Assuming that's accurate, this makes sense, and a comment explaining will be very helpful for future readers.